quickwit/quickwit-search/src/search_permit_provider.rs

```diff
@@ -18,151 +18,312 @@
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
 use std::collections::VecDeque;
-use std::sync::{Arc, Mutex};
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
 
+use bytesize::ByteSize;
 use quickwit_common::metrics::GaugeGuard;
-use tokio::sync::oneshot;
+use quickwit_proto::search::SplitIdAndFooterOffsets;
+#[cfg(test)]
+use tokio::sync::watch;
+use tokio::sync::{mpsc, oneshot};
 
-/// `SearchPermitProvider` is a distributor of permits to perform single split
-/// search operation.
+/// Distributor of permits to perform split search operation.
 ///
-/// Requests are served in order.
+/// Requests are served in order. Each permit initially reserves a slot for the
+/// warmup (limit concurrent downloads) and a pessimistic amount of memory. Once
+/// the warmup is completed, the actual memory usage is set and the warmup slot
+/// is released. Once the search is completed and the permit is dropped, the
+/// remaining memory is also released.
 #[derive(Clone)]
 pub struct SearchPermitProvider {
-    inner_arc: Arc<Mutex<InnerSearchPermitProvider>>,
+    message_sender: mpsc::UnboundedSender<SearchPermitMessage>,
+    #[cfg(test)]
+    actor_stopped: watch::Receiver<bool>,
+}
+
+#[derive(Debug)]
+pub enum SearchPermitMessage {
+    Request {
+        permit_sender: oneshot::Sender<Vec<SearchPermitFuture>>,
+        permit_sizes: Vec<u64>,
+    },
+    UpdateMemory {
+        memory_delta: i64,
+    },
+    FreeWarmupSlot,
+    Drop {
+        memory_size: u64,
+        warmup_slot_freed: bool,
+    },
+}
+
+/// Makes very pessimistic estimate of the memory allocation required for a split search
+///
+/// This is refined later on when more data is available about the split.
+pub fn compute_initial_memory_allocation(
+    split: &SplitIdAndFooterOffsets,
+    warmup_single_split_initial_allocation: ByteSize,
+) -> ByteSize {
+    let split_size = split.split_footer_start;
+    let proportional_allocation =
+        warmup_single_split_initial_allocation.as_u64() * split.num_docs / 10_000_000;
+    let size_bytes = [
+        split_size,
+        proportional_allocation,
+        warmup_single_split_initial_allocation.as_u64(),
+    ]
+    .into_iter()
+    .min()
+    .unwrap();
+    ByteSize(size_bytes)
 }
```

Comment:
I'm not convinced by this estimate logic. The `proportional_allocation` makes it very hard to define what `warmup_single_split_initial_allocation` actually is.

Reply:
Ran some tests on the [hdfs dataset](https://raw.githubusercontent.com/quickwit-oss/quickwit/main/config/tutorials/hdfs-logs/index-config.yaml), and running a query with 3 aggregrations (severity, tenant_id and date) I get a warmup cache in the 50-100MB range already. It feels like the 1GB/10 million docs can be easily exceeded on datasets a tiny bit more pathological (which is probably the case when OOMs happen currently).

Reply:
> What was the size of the split to get this warmup cache size?

Merged splits (~10 million docs) are ~750MB (with the store). A fast field with that many documents can be easily in the 50MB range if it has a poor compression ratio. I don't know the most extreme use cases for the other indexes.